extendeddatacrossword.go

```diff
@@ -73,172 +71,183 @@ func RepairExtendedDataSquare(
-								eds.setCell(uint(p), i, s)
+								eds.setCell(uint(p), uint(i), s)
 							}
 						}
-
-						// Rebuild extended part if incomplete
-						if !vecSliceIsTrue(vectorMask, int(eds.originalDataWidth), int(eds.width)) {
-							if mode == row {
-								rebuiltExtendedShares, err = Encode(eds.rowSlice(i, 0, eds.originalDataWidth), eds.codec)
-							} else if mode == column {
-								rebuiltExtendedShares, err = Encode(eds.columnSlice(0, i, eds.originalDataWidth), eds.codec)
-							}
+						if isExtendedPartIncomplete {
+							err := eds.rebuildExtendedPart(mode, uint(i))
 							if err != nil {
 								return err
 							}
-							for p, s := range rebuiltExtendedShares {
-								if mode == row {
-									eds.setCell(i, eds.originalDataWidth+uint(p), s)
-								} else if mode == column {
-									eds.setCell(eds.originalDataWidth+uint(p), i, s)
-								}
-							}
 						}
 
 						// Check that rebuilt vector matches given merkle root
-						if mode == row {
-							if !bytes.Equal(eds.RowRoot(i), rowRoots[i]) {
-								return &ErrByzantineRow{i}
-							}
-						} else if mode == column {
-							if !bytes.Equal(eds.ColRoot(i), columnRoots[i]) {
-								return &ErrByzantineColumn{i}
-							}
+						err := eds.verifyRoots(rowRoots, columnRoots, mode, uint(i))
+						if err != nil {
+							return err
 						}
 
 						// Check that newly completed orthogonal vectors match their new merkle roots
-						for j := uint(0); j < eds.width; j++ {
-							if vectorMask.AtVec(int(j)) == 0 {
-								if mode == row {
-									adjMask := mask.ColView(int(j))
-									if vecNumTrue(adjMask) == adjMask.Len()-1 && !bytes.Equal(eds.ColRoot(j), columnRoots[j]) {
-										return &ErrByzantineColumn{j}
-									}
-								} else if mode == column {
-									adjMask := mask.RowView(int(j))
-									if vecNumTrue(adjMask) == adjMask.Len()-1 && !bytes.Equal(eds.RowRoot(j), rowRoots[j]) {
-										return &ErrByzantineRow{j}
-									}
+						for j := 0; j < int(eds.width); j++ {
+							if mode == row && !bitMask.Get(i, j) {
+								if bitMask.ColumnIsOne(j) && !bytes.Equal(eds.ColRoot(uint(j)), columnRoots[j]) {
+									return &ErrByzantineColumn{uint(j)}
+								}
+							} else if mode == column && !bitMask.Get(j, i) {
+								if bitMask.RowIsOne(j) && !bytes.Equal(eds.RowRoot(uint(j)), rowRoots[j]) {
+									return &ErrByzantineRow{uint(j)}
 								}
 							}
 						}
 
 						// Set vector mask to true
 						if mode == row {
 							for j := 0; j < int(eds.width); j++ {
-								mask.Set(int(i), j, 1)
+								bitMask.Set(i, j)
 							}
 						} else if mode == column {
 							for j := 0; j < int(eds.width); j++ {
-								mask.Set(j, int(i), 1)
+								bitMask.Set(j, i)
 							}
 						}
-					} else { // repair unsuccessful
-						solved = false
 					}
 				}
 			}
 		}
 
-		if solved {
-			break
-		} else if !progressMade {
+		if !progressMade {
 			return ErrUnrepairableDataSquare
 		}
```

Comment:
Ahh I think this is issue for the failing lazyledger-core test mentioned in the other comment, the square is being repaired fine, but an error is being returned.

Reply:
Good catch!! I'll revert these changes in the for-loop here. They are orthogonal to the bit-mask anyways.

Reply:
Done in https://github.com/lazyledger/rsmt2d/pull/33/commits/3eca5806440382bbe899fe1d66700fedc9c6c2aa.
I'll retest with the above test to make sure that this fixed the issue.

Reply:
An alternative could be to change that if-else-if to: `if !progressMade && !solved` (or `!(progressMade || solved)`) and make the for loop `for !solved`. That should be equivalent. I think the for-loop would be easier to grasp directly (repeat until solved).

Reply:
Either way, thanks for finding that issue and for pointing me to the solution even :-)